airbyte-integrations/connectors/source-youtube-analytics/source_youtube_analytics/source.py

```diff
@@ -0,0 +1,183 @@
+from typing import Any, Iterable, List, Mapping, MutableMapping, Optional, Tuple
+
+import requests
+from airbyte_cdk.sources import AbstractSource
+from airbyte_cdk.sources.streams import Stream
+from airbyte_cdk.sources.streams.http import HttpStream, HttpSubStream
+from airbyte_cdk.sources.streams.http.requests_native_auth import Oauth2Authenticator
+from airbyte_cdk.sources.utils.transform import TransformConfig, TypeTransformer
+
+
+class JobsStream(HttpStream):
+    " https://developers.google.com/youtube/reporting/v1/reference/rest/v1/jobs "
+
+    name = None
+    primary_key = None
+    http_method = None
+    url_base = "https://youtubereporting.googleapis.com/v1/"
+    JOB_NAME = "Airbyte reporting job"
+
+    def next_page_token(self, response: requests.Response) -> Optional[Mapping[str, Any]]:
+        return None
+
+    def parse_response(self, response: requests.Response, **kwargs) -> Iterable[Mapping]:
+        return [response.json()]
+
+    def path(self, **kwargs) -> str:
+        return "jobs"
+
+    def request_body_json(self, **kwargs) -> Optional[Mapping]:
+        if self.name:
+            return {"name": self.JOB_NAME, "reportTypeId": self.name}
+
+    def list(self):
+        " https://developers.google.com/youtube/reporting/v1/reference/rest/v1/jobs/list "
+        self.name = None
+        self.http_method = "GET"
+        results = list(self.read_records(sync_mode=None))
+        result = results[0]
+        return result.get("jobs", {})
+
+    def create(self, name):
+        " https://developers.google.com/youtube/reporting/v1/reference/rest/v1/jobs/create "
+        self.name = name
+        self.http_method = "POST"
+        results = list(self.read_records(sync_mode=None))
+        result = results[0]
+        return result["id"]
+
+
+class ReportResources(HttpStream):
+    " https://developers.google.com/youtube/reporting/v1/reference/rest/v1/jobs.reports/list "
+
+    name = None
+    primary_key = "id"
+    url_base = "https://youtubereporting.googleapis.com/v1/"
+
+    def __init__(self, name: str, jobs_stream: JobsStream, job_id: str, **kwargs):
+        self.name = name
+        self.jobs_stream = jobs_stream
+        self.job_id = job_id
+        return super().__init__(**kwargs)
+
+    def next_page_token(self, response: requests.Response) -> Optional[Mapping[str, Any]]:
+        return None
+
+    def parse_response(self, response: requests.Response, **kwargs) -> Iterable[Mapping]:
+        response_json = response.json()
+        reports = []
+        for report in response_json.get("reports", []):
+            report = {**report}
+            report["startTime"] = datetime.datetime.strptime(report["startTime"], "%Y-%m-%dT%H:%M:%S%z")
+            reports.append(report)
+        reports.sort(key=lambda x: x["startTime"])
+        date = kwargs["stream_state"].get("date")
+        if date:
+            reports = [r for r in reports if int(r["startTime"].date().strftime("%Y%m%d")) >= date]
```

Comment:
this filter should be done on API side
https://developers.google.com/youtube/reporting/v1/reference/rest/v1/jobs.reports/list

Reply:
Parameters:
    createdAfter

Reply:
Yes I know about this query param, but it's not what I need

This parameter filter on report `createTime` field but for main logic I need to use `startTime` field.
Report `startTime` and `createTime` are different things,
report can cover more earlier period than it was created.

Also YouTube don't  keep reports infinitely.
It will start to remove old reports,  number of reports will always be about 60 - 90.